File: Programmers/Level_3/스타수열.py
'''
https://programmers.co.kr/learn/courses/30/lessons/70130

간과한 점: 
수열을 무조건 n-1번째와 n번째가 붙은 상태로 자르는게 아닐 수도 있다
즉, n번째와 n+4번째가 하나의 뭉치로 될 수 있다

풀이:
a배열의 각 원소 개수를 센다
만약 특정 원소의 개수 * 2가 지금까지 구한 값 보다 작다면 최댓값이 될 수 없다는 뜻이니 건너뛴다

n과 n+1이 같으면 조건에 위배되고, 
특정 원소가 n과 n+1 둘 중 어느 것에도 해당하지 않는다면 공통 원소가 없어서 조건에 위배되니
인덱스를 1만 증가시키고 건너뛴다 (n+1, n+2가 조건에 부합하는지 체크해야하니까)

만약 조건에 위배되는 사항이 없다면 현재의 특정 원소로 구할 수 있는 부분 수열 길이를 2 증가시키고 인덱스도 2 증가시킨다
최종 부분 수열의 길이가 가장 큰 것을 리턴한다
'''

from collections import Counter

def solution(a):
    if len(a) < 2:
        return 0
    
    answer = 0
    counter = Counter(a)
    
    for key, value in counter.items():
        if value * 2 < answer:
            continue
        
        idx, l = 0, 0
        while idx < len(a)-1:
            if (key not in a[idx:idx+2]) or a[idx] == a[idx+1]:
                idx += 1
                continue
            
            idx, l = idx + 2, l + 2
            
        answer = max(answer, l)
        
    return answer

File: Programmers/Level_3/Swift_Code/Swift_Code/스타수열.swift
//
//  스타수열.swift
//  Swift_Code
//
//  Created by InJe Choi on 2022/03/30.
//

/*
 https://programmers.co.kr/learn/courses/30/lessons/70130?language=swift
 
 배열이 크든 작든 Array Slicing 후 contains 메소드를 통해 조사하는 방식은 시간 효율면에서 좋지 않다
 Array Slice는 결국 원본 배열의 복제본을 하나 만들고 참조하고 있는 것이기 때문에
 이 작업 자체가 무겁다고 볼 수 있다
 여기서 또 contains 메소드를 통해 O(N)의 시간 복잡도로 조사를 하기 때문에 시간 효율면에서 매우 안 좋다
 
 고로, 배열의 특정 범위에서만 조회하고 싶어도 Subsequence를 따로 만들지 말고
 웬만해선 원본 배열에서 특정 범위를 반복문을 통해 순회하면서 값을 직접 비교하는 것이 더욱 빠르다
 
 ex)
 지양해야 하는 방식: arr[idx...idx+10].contains($0)
 */

import Foundation

func solution(_ a: [Int]) -> Int {
    if a.count < 2 {
        return 0
    }
    
    var answer = 0
    var counter = [Int: Int]()
    a.forEach { counter[$0, default: 0] += 1 }
    
    counter.forEach {
        if $1 * 2 < answer {
            return
        }
        var idx = 0, length = 0
        while idx < a.count - 1 {
            if (a[idx] != $0 && a[idx+1] != $0) || a[idx] == a[idx+1] {
                idx += 1
                continue
            }
            idx += 2
            length += 2
        }
        answer = max(answer, length)
    }
    
    return answer
}

// print(solution([5,2,3,3,5,3]))
